### PR TITLE
Add Multiple LDM Compatible Samplers 

### PR DESCRIPTION
**Pull request by ExponentialML**
Here's a feature that allows for multiple samplers to be used. A dropdown is added to the extension with two supported samplers, `DDIM, UniPC`. DPM Solver may also be supported as it's apart of the LDM code base. This will also sunset the current DDIM sampler, or make it a legacy one.

The sampler code has been refactored so that we can use and modify existing ones. Originally the plan was to leverage the built-in samplers in the webui, but there would be too many hooks and workarounds to make it work properly.

Each sampler is modified in some way to allow for separable conditioning (conditioning, unconditioning) rather than chunks, and the UniPC has some modifications that allows it to work properly. Currently, there is a  bug that is preventing DDIM from working properly, but vid2vid should work just fine.

Before merging, the following must be resolved.

- [x] Fix the bug that doesn't allow sampling from noise when using DDIM.
- [x] Refactor code and remove comments.
- [x] Fully deprecate (or make it legacy) the DDIM sampler on the UNETSD model.
- [x] Ensure that the inpainting masks work.
- [x] Update code to track with any pushed future PRs.

After that, we should be good to go :+1: .

### CODE REVIEW

**Comment by ExponentialML** [comment]
This is now ready for testing and review.

image

**Comment by B34STW4RS** [comment]
Works pretty good, however I noticed it seems to massively change vid2vid content relative to the original sampler. For instant ddim_gaussian gave me a white space marine looking to the right and firing a rifle, while the original sampler gave me a blue space marine firing a laser pistol looking left.

https://github.com/kabachuha/sd-webui-text2video/assets/11381013/6458a912-5544-48e6-b39b-c7fe9a0c56d8

Unipc seems to give the expected result howerver, eg:



https://github.com/kabachuha/sd-webui-text2video/assets/11381013/37c11328-7a88-47d3-8504-65959e2c6252

Overall good work will keep testing over the weekend.

File: scripts/modelscope/t2v_model.py [hunk 1]
```diff
@@ -488,18 +552,18 @@ def forward(self, x, context=None, mask=None):
             mask = rearrange(mask, 'b ... -> b (...)')
             max_neg_value = -torch.finfo(x.dtype).max
             mask = repeat(mask, 'b j -> (b h) () j', h=h)
-
-        if getattr(cmd_opts, "force_enable_xformers", False) or (getattr(cmd_opts, "xformers", False) and shared.xformers_available and torch.version.cuda and (6, 0) <= torch.cuda.get_device_capability(shared.device) <= (9, 0)):
+        
+        if cmd_opts.force_enable_xformers or (cmd_opts.xformers and shared.xformers_available and torch.version.cuda and (6, 0) <= torch.cuda.get_device_capability(shared.device) <= (9, 0)):
```
Comment (kabachuha) [inline scripts/modelscope/t2v_model.py:556]: You caused an accidental regression of the Vlad-compatibility fix

Comment (kabachuha) [reply scripts/modelscope/t2v_model.py:556]: I'll edit it out (if I can edit the PR)

File: scripts/modelscope/t2v_pipeline.py [hunk 2]
```diff
@@ -18,14 +18,17 @@
 import torch.cuda.amp as amp
 from einops import rearrange
 import cv2
-from modelscope.t2v_model import UNetSD, AutoencoderKL, GaussianDiffusion, beta_schedule
-from modules import devices
+from scripts.modelscope.t2v_model import UNetSD, AutoencoderKL, GaussianDiffusion, beta_schedule
```
Comment (kabachuha) [inline scripts/modelscope/t2v_pipeline.py:21]: Again, merger fix

File: scripts/modelscope/t2v_pipeline.py [hunk 3]
```diff
@@ -192,7 +193,26 @@ def compute_latents(self, vd_out, cpu_vae='GPU (half precision)', device=torch.d
...
+        eta=0.0, 
+        cpu_vae='GPU (half precision)', 
+        device=torch.device('cpu'), 
+        latents=None, 
+        skip_steps=0,
+        strength=0,
+        mask=None, 
+        is_vid2vid=False,
+        sampler=available_samplers[0].name
+    ):
```
Comment (kabachuha) [inline scripts/modelscope/t2v_pipeline.py:215]: Good idea to expand the definitions, it will be easier to track changes this way

**Review by kabachuha — APPROVED** [review]
Everything works fine now

Tested all the three samplers
